src/crypto-providers/util.ts

```diff
@@ -0,0 +1,262 @@
+  addressToBuffer,
+  getBootstrapAddressProtocolMagic,
+} = require('cardano-crypto.js')
+
+const isShelleyPath = (path: number[]) => path[0] - HARDENED_THRESHOLD === 1852
+
+const isStakingPath = (path: number[]) => path[3] === 2
+
+const encodeAddress = (address: Buffer): string => {
+  const addressType = getAddressType(address)
+  if (addressType === AddressTypes.BOOTSTRAP) {
+    return base58.encode(address)
+  }
+  const addressPrefixes: {[key: number]: string} = {
+    [AddressTypes.BASE]: 'addr',
+    [AddressTypes.POINTER]: 'addr',
+    [AddressTypes.ENTERPRISE]: 'addr',
+    [AddressTypes.REWARD]: 'stake',
+  }
+  const isTestnet = getShelleyAddressNetworkId(address) === NetworkIds.TESTNET
+  const addressPrefix = `${addressPrefixes[addressType]}${isTestnet ? '_test' : ''}`
+  return bech32.encode(addressPrefix, address)
+}
+
+const getSigningPath = (
+  signingFiles: HwSigningData[], i: number,
+): BIP32Path | undefined => {
+  if (!signingFiles.length) return undefined
+  return signingFiles[i] ? signingFiles[i].path : signingFiles[0].path
+}
+
+const filterSigningFiles = (
+  signingFiles: HwSigningData[],
+): {paymentSigningFiles: HwSigningData[], stakeSigningFiles: HwSigningData[]} => {
+  const paymentSigningFiles = signingFiles.filter(
+    (signingFile) => signingFile.type === 0,
+  )
+  const stakeSigningFiles = signingFiles.filter(
+    (signingFile) => signingFile.type === 1,
+  )
+  return {
+    paymentSigningFiles,
+    stakeSigningFiles,
+  }
+}
+
+const findSigningPath = (
+  certPubKeyHash: Buffer, stakingSigningFiles: HwSigningData[],
+): BIP32Path | undefined => {
+  const signingFile = stakingSigningFiles.find((file) => {
+    const { pubKey } = XPubKey(file.cborXPubKeyHex)
+    const pubKeyHash = getPubKeyBlake2b224Hash(pubKey)
+    return !Buffer.compare(pubKeyHash, certPubKeyHash)
+  })
+  return signingFile?.path
+}
+
+const txHasPoolPoolRegistrationCert = (
+  certs: _Certificate[],
+): boolean => certs.some(
+  ({ type }) => type === TxCertificateKeys.STAKEPOOL_REGISTRATION,
+)
+
+const validateTx = (
+  txAux: _TxAux, paymentSigningFiles: HwSigningData[], stakeSigningFiles: HwSigningData[],
+): void => {
+  if (!txAux.inputs.length) throw NamedError('MissingInputError')
+  if (!txAux.outputs.length) throw NamedError('MissingOutputError')
+  if (paymentSigningFiles.length > txAux.inputs.length) {
+    throw NamedError('TooManySigningFilesError')
+  }
+  const requireStakingSigningFile = !!(txAux.certificates.length + txAux.withdrawals.length)
+  if (
+    requireStakingSigningFile && !stakeSigningFiles.length
+  ) throw NamedError('MissingStakingSigningFileError')
+}
+
+const validateWitnessing = (
+  txAux: _TxAux, signingFiles: HwSigningData[],
+): void => {
+  const {
+    paymentSigningFiles,
+    stakeSigningFiles,
+  } = filterSigningFiles(signingFiles)
+  validateTx(txAux, paymentSigningFiles, stakeSigningFiles)
+  if (!txHasPoolPoolRegistrationCert(txAux.certificates)) return
```

Comment:
"PoolPoolRegistration" seems to me like a typo/incorrect find-replace attempt

Reply:
also, I would consider more transparent to have the "ifs" below encapsulated by a validateStakepoolRegistration conditions or something, as this return statement is easy to overlook among all those ifs

Reply:
:D fixed thanks.